src/common/friendly_printer.h

```diff
@@ -0,0 +1,196 @@
+/// associated stream. Consider gating expensive evaluations in
+/// critical code behind this.
+///
+/// We annotate the printing case as unlikely in all conditions since
+/// we're most likely to be run as part of a daemon.
+///
+/// Properties
+/// ----------
+///
+/// Setting a property returns a new `friendly_printer` with that
+/// property set. This is so we can pass one down into classes
+/// operating on different parts of the work and set values as
+/// appropriate. An rvalue overload of with_prop will skip the copy.
+///
+/// To use a property, we provide a user literal.
+///
+/// Examples
+/// --------
+///
+/// ```
+/// friendly_printer fp{&std::cout}
+///
+/// catastrophize(5, fp.with_prop("doom", "exploded"));
+///
+/// void catastrophize(int n, const friendly_print& fp) {
+///   fp.print("Oh, no! After being hit with {} {}, the world has {}!",
+///            n ? 1 "comet" : "comets", "doom"_fpk);
+/// }
+/// ```
+class friendly_printer {
+public: // Not really, but literal operators don't work as friend
+	// functions
+
+  /// This class is a wrapper around `std::string_view`, disjoint so
+  /// our magic overload of `sub` can replace it with the actual value
+  /// from a property.
+  class magic_sub_ {
+  public:
+    std::string_view k;
+    magic_sub_(const char* s, std::size_t l)
+      : k(s, l) {}
+  };
+
+private:
+
+  std::ostream* const ostr{nullptr};
+  // Since heterogeneous lookup for `unordered_map` isn't until C++20
+  boost::container::flat_map<std::string, std::string, std::less<>> props;
+
+  /// To simplify our template code, pass everything through sub and
+  /// make it a no-op for everything that isn't `magic_sub_`
+  template<typename T>
+  decltype(auto) sub(T&& t,
+		     std::enable_if_t<!std::is_same_v<
+		       std::decay_t<T>, magic_sub_>>* = nullptr) const {
+    return std::forward<T>(t);
+  }
+
+  /// Substitute in a property value. We don't need to gate here,
+  /// because we're only called from print.
+  std::string_view sub(const magic_sub_& s) const {
+    using namespace std::literals;
+    if (auto i = props.find(s.k); likely(i != props.cend())) {
+      return i->second;
+    } else {
+      return "<UNFOUND>"sv;
+    }
+  }
+
+  /// If we know how to turn things into strings, just let them pass them in.
+  template<typename T>
+  static std::string stringize(T t) {
+    if constexpr (std::is_convertible_v<T, std::string> ||
+	std::is_same_v<std::decay_t<T>, std::string>) {
+      return t;
+    } else if constexpr (std::is_same_v<std::decay_t<T>, std::string_view>) {
+      return std::string(t);
+    } else {
+      return fmt::format("{}", t);
+    }
+  }
+
+public:
+
+  friendly_printer(std::ostream* ostr)
+    : ostr(ostr) {}
+
+  /// Print message if we have a stream, substituting properties along
+  /// the way.
+  template<typename ...T>
+  void print(T&& ...t) const {
+    if (unlikely(!!ostr)) {
+      fmt::print(*ostr, sub(t)...);
```

Comment:
The main counter-argument in this case is getting the actual name of the zone instead of a zone id.

(That's what made me start down the rabbit hole of substitution.)

Everything *else*, I think should actually be already down there and getable, I think.

Though just fetching the zone name probably isn't all that intensive, if we're printing stuff anyway.

Though I think at some point we'll definitely want something along these lines, specifically a pile of annotations that we can modify as we go. Though that might be more apropos to the tracing framework overall.

Reply:
Done!

Reply:
> Though that might be more apropos to the tracing framework overall.

yeah, tracing is well-suited for that. composition of `DoutPrefixProvider`s can do that for logging as well